### PR TITLE
Add ssh-port argument

### PR DESCRIPTION
**Pull request by dlrobertson**
Add the ssh-port option and find an open port if there is no port given
instead of using a default value.

### CODE REVIEW

**Comment by ALSchwalm** [comment]
:joy: I did pretty much the same in https://github.com/ALSchwalm/transient/commit/70d2adc030bd03d7ef97ff0ab705c7c9b1e49bcd

Maybe we can use the best of both worlds.

File: transient/transient.py [hunk 1]
```diff
@@ -30,18 +31,33 @@ def __format_new_qemu_args(self) -> List[str]:
             # Use userspace networking (so no root is needed), and bind
             # localhost port 5555 to guest port 22
             # TODO: try to find an available port
-            new_args.extend(["-net", "nic,model=e1000",
-                             "-net", "user,hostfwd=tcp::5555-:22"])
+            new_args.extend([
+                "-netdev",
+                "user,id=sshdev0,hostfwd=tcp::{}-:22".format(self.config["ssh_port"]),
+                "-device",
+                "virtio-net-pci,netdev=sshdev0"
```
Comment (ALSchwalm) [inline transient/transient.py]: what is the reason for the change here? does this not now require the guest have virtio support?

Comment (dlrobertson) [reply transient/transient.py]: `virtio` isn't needed, but `-net` is deprecated in favor of `-netdev`

**Comment by dlrobertson** [comment]
> :joy: I did pretty much the same in 70d2adc
>
> Maybe we can use the best of both worlds.

ha! probably should have checked the dev branch.

File: transient/transient.py (hunk 1 above)
Comment (ALSchwalm) [reply transient/transient.py]: Gotcha. It might be good to leave the older approach if it means we're compatible with more stuff. I guess we should pick a minimum qemu version. Do you have a recommendation? Maybe whatever is in the Ubuntu 18.04 repos?

Comment (dlrobertson) [reply transient/transient.py]: AFAIK ubuntu 18.04 is on 2.11 and `-netdev` was introduced before that.

**Comment by ALSchwalm** [comment]
I had to do some restructuring to the ssh logic to support sshfs. I'd still like to pull in the user-specified `ssh-port`  support and `netdev` usage from this (provided we can use `netdev` without requiring guest virtio support). Are you interested in rebasing this?

**Comment by dlrobertson** [comment]
Yeah I can do that

**Comment by dlrobertson** [comment]
Updated

File: transient/transient.py [hunk 2]
```diff
@@ -41,11 +41,19 @@ def __qemu_added_devices(self) -> List[str]:
...
-            new_args.extend(["-net", "nic,model=e1000",
-                             "-net", "user,hostfwd=tcp::{}-:22".format(self.ssh_port)])
+            if not self.config.ssh_port:
+                self.ssh_port = self.__allocate_random_port()
+            else:
+                self.ssh_port = self.config.ssh_port
+
+            new_args.extend([
+                "-netdev",
+                "user,id=sshdev0,hostfwd=tcp::{}-:22".format(self.ssh_port),
```
Comment (ALSchwalm) [inline transient/transient.py]: Can we set an ID that contains 'transient'? Just to ensure we don't run in to anything a user normally does.

File: transient/transient.py [hunk 3]
```diff
@@ -41,11 +41,19 @@ def __qemu_added_devices(self) -> List[str]:
             if self.config.ssh_console is True:
                 new_args.append("-nographic")
 
-            # Use userspace networking (so no root is needed), and bind
             # the random localhost port to guest port 22
-            self.ssh_port = self.__allocate_random_port()
-            new_args.extend(["-net", "nic,model=e1000",
-                             "-net", "user,hostfwd=tcp::{}-:22".format(self.ssh_port)])
+            if not self.config.ssh_port:
```
Comment (ALSchwalm) [inline transient/transient.py]: I would use `if self.config.ssh_port is None` here.

Comment (dlrobertson) [reply transient/transient.py]: Good call

File: transient/transient.py (hunk 2 above)
Comment (dlrobertson) [reply transient/transient.py]: yup! updated

**Comment by ALSchwalm** [comment]
Awesome. Thanks for working on this, sorry about the conflict.

**Comment by dlrobertson** [comment]
np man
